src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/IdentifiersShouldNotMatchKeywordsTypeRuleTests.cs

```diff
@@ -0,0 +1,124 @@
+    /// </summary>
+    public partial class IdentifiersShouldNotMatchKeywordsTests
+    {
+        [Fact]
+        public void CSharpDiagnosticForKeywordNamedPublicType()
+        {
+            VerifyCSharp(@"
+public class @class {}
+",
+                GetCSharpResultAt(2, 14, IdentifiersShouldNotMatchKeywordsAnalyzer.TypeRule, "class", "class"));
+        }
+
+        [Fact]
+        public void BasicDiagnosticForKeywordNamedPublicType()
+        {
+            VerifyBasic(@"
+Public Class [Class]
+End Class
+",
+                GetBasicResultAt(2, 14, IdentifiersShouldNotMatchKeywordsAnalyzer.TypeRule, "Class", "Class"));
+        }
+
+        [Fact]
+        public void CSharpNoDiagnosticForCaseSensitiveKeywordNamedPublicTypeWithDifferentCasing()
+        {
+            VerifyCSharp(@"
+public class iNtErNaL {}
+");
+        }
+
+        [Fact]
+        public void BasicNoDiagnosticForCaseSensitiveKeywordNamedPublicTypeWithDifferentCasing()
+        {
+            VerifyBasic(@"
+Public Class iNtErNaL
+End Class");
+        }
+
+        [Fact]
+        public void CSharpDiagnosticForCaseInsensitiveKeywordNamedPublicType()
+        {
+            VerifyCSharp(@"
+public struct aDdHaNdLeR {}
+",
+                GetCSharpResultAt(2, 15, IdentifiersShouldNotMatchKeywordsAnalyzer.TypeRule, "aDdHaNdLeR", "AddHandler"));
+        }
+
+        [Fact]
+        public void BasicDiagnosticForCaseInsensitiveKeywordNamedPublicType()
+        {
+            VerifyBasic(@"
+Public Structure aDdHaNdLeR
+End Structure",
+                GetBasicResultAt(2, 18, IdentifiersShouldNotMatchKeywordsAnalyzer.TypeRule, "aDdHaNdLeR", "AddHandler"));
+        }
+
+        [Fact]
+        public void CSharpNoDiagnosticForKeywordNamedInternalype()
+        {
+            VerifyCSharp(@"
+internal class @class {}
+");
+        }
+
+        [Fact]
+        public void BasicNoDiagnosticForKeywordNamedInternalType()
+        {
+            VerifyBasic(@"
+Friend Class [Class]
+End Class
+");
+        }
+
+        [Fact]
+        public void CSharpNoDiagnosticForNonKeywordNamedPublicType()
+        {
+            VerifyCSharp(@"
+public class classic {}
+");
+        }
+
+        [Fact]
+        public void BasicNoDiagnosticForNonKeywordNamedPublicType()
+        {
+            VerifyBasic(@"
+Public Class Classic
+End Class
+");
+        }
+
+        [Fact]
+        public void CSharpDiagnosticForKeywordNamedPublicTypeInNamespace()
+        {
+            VerifyCSharp(@"
+namespace N
+{
+    public enum @enum {}
+}
+",
+                GetCSharpResultAt(4, 17, IdentifiersShouldNotMatchKeywordsAnalyzer.TypeRule, "enum", "enum"));
+        }
+
+        [Fact]
+        public void BasicDiagnosticForKeywordNamedPublicTypeInNamespace()
+        {
+            VerifyBasic(@"
+Namespace N
+    Public Enum [Enum]
+    End Class
+End Namespace
+",
+                GetBasicResultAt(3, 17, IdentifiersShouldNotMatchKeywordsAnalyzer.TypeRule, "Enum", "Enum"));
+        }
+    }
+}
```

Comment:
As we discussed: your idea here is to make sure that nobody "cleans up" my usage of `GetResultantVisibility` and changes it to `DeclaredAccessibility`. I will do this in the next PR, the last one, where I implement the namespace rule.